stable-diffusion/data.py

```diff
@@ -0,0 +1,132 @@
+        transforms.RandomHorizontalFlip()
+        if random_flip else transforms.Lambda(lambda x: x),
+        transforms.ToTensor(),
+        transforms.Normalize(mean, std),
+    ])
+
+    def tokenize_captions(examples: dict,
+                          tokenizer: callable = tokenizer,
+                          caption_column: str = caption_column,
+                          is_train: bool = True):
+        captions = []
+        for caption in examples[caption_column]:
+            if isinstance(caption, str):
+                captions.append(caption)
+            elif isinstance(caption, (list, np.ndarray)):
+                # take a random caption if there are multiple
+                captions.append(
+                    random.choice(caption) if is_train else caption[0])
+            else:
+                raise ValueError(
+                    f"Caption column `{caption_column}` should contain either strings or lists of strings."
+                )
+        inputs = tokenizer(captions,
+                           max_length=tokenizer.model_max_length,
+                           padding="max_length",
+                           truncation=True,
+                           return_tensors="pt")
+        return inputs.input_ids
+
+    def preprocess(examples: dict):
+        images = [image.convert("RGB") for image in examples[image_column]]
+        examples["image_tensor"] = [train_transforms(image) for image in images]
+        examples["input_ids"] = tokenize_captions(examples)
+        return examples
+
+    with dist.run_local_rank_zero_first():
+        dataset = load_dataset(name, split='train')
+
+    # add image_tensor and input_ids columns (processed images and text)
+    dataset = dataset.with_transform(preprocess)
+    sampler = dist.get_sampler(dataset, drop_last=drop_last, shuffle=shuffle)
+    return DataSpec(dataloader=DataLoader(dataset=dataset,
+                                          batch_size=batch_size,
+                                          sampler=sampler,
+                                          drop_last=drop_last,
+                                          collate_fn=collate_fn,
+                                          **dataloader_kwargs))
+
+
+def build_prompt_dataspec(prompts: list, batch_size: int, **dataloader_kwargs):
+    """prompt dataset for eval"""
+    dataset = PromptDataset(prompts)
+    sampler = dist.get_sampler(dataset, drop_last=False, shuffle=False)
+    
+    def split_list(l, microbatch_size: int): # needed for gradient accumulation
+        if len(l) < microbatch_size:
+            microbatch_size = len(l)
+        num_microbatches = math.ceil(len(l) / microbatch_size)
+        # Note: this is to match the behavior of tensor.chunk, which is used in _split_tensor
+        chunked_microbatch_size = math.ceil(len(l) / num_microbatches)
+        return [l[start:start + chunked_microbatch_size] for start in range(0, len(l), chunked_microbatch_size)]
```

Comment:
Could this be avoided if `generate` took in tokenized text instead of raw strings? The change seems worth it to avoid this extra function

Reply:
Composer is/was being updated to handle strings. I'll test the new version and update accordingly. Generate taking tokens would be clunky to use at inference time imo.

Reply:
For inference, it shouldn't be too bad to add a tokenizer as a preprocessor to input data, right? 

I was motivated to ask since it was taking me a few hours to get COCO captions dataloader to output raw strings and for composer to accept them. It made it way easier when `generate()` could accept tokenized text.